# IFQ718 Assignment 2

---

## Questions

Select any three

1. What is the profit of each business per year?

2. Which product generated the most profit for each business per year?

3. What was the most popular product for each business in each year?

4. Are there periods of the year where some businesses are more profitable?

5. What were the most utilised promotions for each business?

6. Which customers were most loyal for each business?

7. What is the total expenditure of loyal customers compared to one-off customers?

8. What is the employee turnover rate of each business?

9. How much has inflation impacted the profit margin of each business?

10. What is the impact of seasonal fluctuations on each business?

---

## Getting started

We have provided some code below to get you started

In [7]:
!pip install duckdb
!pip install plotly

Looking in indexes: https://pypi.org/simple, https://****@pypi.fury.io/jackptoke/lewagon_jackptoke_gcp_vm/
Looking in indexes: https://pypi.org/simple, https://****@pypi.fury.io/jackptoke/lewagon_jackptoke_gcp_vm/


In [9]:
# Run this cell to begin working with this Notebook
# Please make any modifications to suit your needs

import os
import csv
import json

import pandas as pd
import duckdb

---

# Preparations

## All the data transformation are happening in dbt, which you can see in the `./dbt/airbnb/models` folder

In [14]:
# All the data tranformation are happ
DUCKDB_FILE = './dbt/airbnb/receipts.duckdb'
with duckdb.connect(DUCKDB_FILE) as conn:
    receipts_df = conn.sql("SELECT * FROM receipts.main.receipts").df()
    receipt_businesses_df = conn.sql("SELECT * FROM receipts.main.receipt_businesses").df()
    receipt_cashiers_df = conn.sql("SELECT * FROM receipts.main.receipt_cashiers").df()
    receipt_customers_df = conn.sql("SELECT * FROM receipts.main.receipt_customers").df()
    receipt_payments_df = conn.sql("SELECT * FROM receipts.main.receipt_payments").df()
    receipt_products_df = conn.sql("SELECT * FROM receipts.main.receipt_products").df()
    receipt_promotions_df = conn.sql("SELECT * FROM receipts.main.receipt_promotions").df()


#### All receipts that we have

In [16]:
##### All the receipts
receipts_df

receipt_ref  sequence_num receipt_date    gst  terminal_num  \
0      9626510:290218:03933          3933   2029/02/18   8.44             4   
1      5593822:260423:01227          1227   2026/04/23   3.98             5   
2      2209696:290126:03734          3734   2029/01/26   0.93             4   
3      3795898:250425:01050          1050   2025/04/25   9.86             4   
4      2209696:240130:00749           749   2024/01/30   1.80             1   
...                     ...           ...          ...    ...           ...   
13475  5593822:280425:01948          1948   2028/04/25   4.67             3   
13476  5593822:270506:01591          1591   2027/05/06   4.58             3   
13477  5593822:230727:00320           320   2023/07/27  11.07             4   
13478  9626510:231230:00706           706   2023/12/30   3.69             3   
13479  2209696:260106:01801          1801   2026/01/06   1.86             1   

       receipt_total  
0              92.82  
1              43.80  
2              10.20  
3             108.48  
4              19.80  
...              ...  
13475          51.32  
13476          50.40  
13477         121.80  
13478          40.56  
13479          20.46  

[13480 rows x 6 columns]

#### This table contains data of which receipts belongs to which businesses

In [17]:
# Normalised table containing all the 
receipt_businesses_df

receipt_ref business_abn              business_name
0      9626510:290218:03933    962 65 10  Please Bring Pizza Pronto
1      5593822:260423:01227    559 38 22        Penguin Swim School
2      2209696:290126:03734    220 96 96        Wake Up with Coffee
3      3795898:250425:01050    379 58 98       Ed's Barber Supplies
4      2209696:240130:00749    220 96 96        Wake Up with Coffee
...                     ...          ...                        ...
13475  5593822:280425:01948    559 38 22        Penguin Swim School
13476  5593822:270506:01591    559 38 22        Penguin Swim School
13477  5593822:230727:00320    559 38 22        Penguin Swim School
13478  9626510:231230:00706    962 65 10  Please Bring Pizza Pronto
13479  2209696:260106:01801    220 96 96        Wake Up with Coffee

[13480 rows x 3 columns]

#### This table contains data of which receipts was produced by which Cashiers

In [18]:
receipt_cashiers_df

receipt_ref cashier_name
0      9626510:290218:03933     Jennifer
1      5593822:260423:01227        Donna
2      2209696:290126:03734       Lauren
3      3795898:250425:01050     Jonathon
4      2209696:240130:00749        Scott
...                     ...          ...
13475  5593822:280425:01948        Donna
13476  5593822:270506:01591        Kevin
13477  5593822:230727:00320        Kevin
13478  9626510:231230:00706         Tara
13479  2209696:260106:01801       Leslie

[13480 rows x 2 columns]

#### This table contains data of which receipts belong to which customers

In [19]:
receipt_customers_df

receipt_ref customer_name  customer_points
0      9626510:290218:03933         Scott             9281
1      5593822:260423:01227         Scott             3310
2      2209696:290126:03734         Isaac             1020
3      3795898:250425:01050         Scott            10848
4      2209696:240130:00749          None             <NA>
...                     ...           ...              ...
13475  5593822:280425:01948         Derek             5132
13476  5593822:270506:01591          None             <NA>
13477  5593822:230727:00320         Derek            12180
13478  9626510:231230:00706          Mary             4055
13479  2209696:260106:01801          None             <NA>

[13480 rows x 3 columns]

#### This table contains data of which payments belong to which receipts

In [20]:
receipt_payments_df

receipt_ref  payment_method  payment_amount
0      9626510:290218:03933          EFTPOS           92.82
1      5593822:260423:01227  Loyalty Points           10.70
2      5593822:260423:01227       Gift Card           33.10
3      2209696:290126:03734     Credit Card            1.02
4      2209696:290126:03734      Debit Card            9.18
...                     ...             ...             ...
19702  5593822:230727:00320     Credit Card          121.80
19703  9626510:231230:00706            Cash            3.85
19704  9626510:231230:00706          EFTPOS           36.71
19705  2209696:260106:01801            Cash           19.35
19706  2209696:260106:01801     Credit Card            1.11

[19707 rows x 3 columns]

In [21]:
#### This table contains data of which products belong to which receipt

In [22]:
receipt_products_df

receipt_ref   product_name  product_cost  product_price  \
0      9626510:290218:03933  Spicy chicken           5.8           18.2   
1      9626510:290218:03933      Pepperoni           7.6           18.2   
2      5593822:260423:01227    Aquaerobics           7.8           14.6   
3      2209696:290126:03734          Mocha           3.4            5.1   
4      3795898:250425:01050          Broom          14.2           44.8   
...                     ...            ...           ...            ...   
33771  9626510:231230:00706          Chips           3.2            6.3   
33772  9626510:231230:00706  Soft drink 1L           1.5            3.7   
33773  2209696:260106:01801     Flat white           3.6            5.5   
33774  2209696:260106:01801          Mocha           2.9            4.4   
33775  2209696:260106:01801          Mocha           2.9            4.4   

       product_quantity  
0                     3  
1                     3  
2                     3  
3                     2  
4                     1  
...                 ...  
33771                 2  
33772                 1  
33773                 1  
33774                 1  
33775                 3  

[33776 rows x 5 columns]

#### This table contains data of which promotions belong to which receipts

In [24]:
receipt_promotions_df

receipt_ref   product_name  discount  per_quantity
0     9626510:290218:03933      Pepperoni      0.25             1
1     9626510:290218:03933  Spicy chicken      0.05             3
2     3795898:250425:01050       Scissors      0.20             2
3     3795898:230810:00400      Cape grey      0.15             2
4     3795898:230810:00400     Soap 500mL      0.10             2
...                    ...            ...       ...           ...
7151  2209696:280728:03391     Long Black      0.30             1
7152  3795898:260213:01378     Cape white      0.25             1
7153  5593822:280425:01948    Aquaerobics      0.20             3
7154  9626510:231230:00706          Chips      0.05             3
7155  2209696:260106:01801          Mocha      0.15             1

[7156 rows x 4 columns]

---

## Exploration One
### Are there periods of the year where some businesses are more profitable?

### Introduction

To answer the above questions, we need to:
1. First, figure out how much sales does each business made.
2. Second, calculate how much profit do they make on each sales after deducting all the discounts, if there are any.
3. Third, we aggregate the profits by the month, the transactions occurred.
4. Finally, we will see if we can see any pattern and then answer the above questions.

### Methods

In [25]:
# Type your method in this cell
with duckdb.connect(DUCKDB_FILE) as conn:
    business_sales_df = conn.sql("SELECT * FROM receipts.main.business_sales").df()
    sale_profits_df = conn.sql("SELECT * FROM receipts.main.sale_profits").df()
    business_sales_profits_df = conn.sql("SELECT * FROM receipts.main.business_sales_profits").df()

### 1. Sales by Businesses - All sales made by the businesses
Since we have normalised the tables, first we need to see which receipts belong to which businesses.

In [28]:
business_sales_df

business_abn              business_name           receipt_ref  \
0        220 96 96        Wake Up with Coffee  2209696:220701:00002   
1        220 96 96        Wake Up with Coffee  2209696:220701:00003   
2        220 96 96        Wake Up with Coffee  2209696:220701:00001   
3        220 96 96        Wake Up with Coffee  2209696:220702:00006   
4        220 96 96        Wake Up with Coffee  2209696:220702:00007   
...            ...                        ...                   ...   
13475    962 65 10  Please Bring Pizza Pronto  9626510:290628:04185   
13476    962 65 10  Please Bring Pizza Pronto  9626510:290628:04186   
13477    962 65 10  Please Bring Pizza Pronto  9626510:290629:04188   
13478    962 65 10  Please Bring Pizza Pronto  9626510:290629:04187   
13479    962 65 10  Please Bring Pizza Pronto  9626510:290629:04189   

      receipt_date  receipt_total    gst  year  month month_name  
0       2022/07/01          60.50   5.50  2022      7       July  
1       2022/07/01          88.50   8.05  2022      7       July  
2       2022/07/01          36.00   3.27  2022      7       July  
3       2022/07/02         112.00  10.18  2022      7       July  
4       2022/07/02          33.00   3.00  2022      7       July  
...            ...            ...    ...   ...    ...        ...  
13475   2029/06/28          92.85   8.44  2029      6       June  
13476   2029/06/28         151.47  13.77  2029      6       June  
13477   2029/06/29          90.20   8.20  2029      6       June  
13478   2029/06/29          33.20   3.02  2029      6       June  
13479   2029/06/29          66.00   6.00  2029      6       June  

[13480 rows x 9 columns]

### 2. Sale Profits - Profit made from each sale
Second we have to calculate how much profits are made by each receipts.

In [29]:
sale_profits_df

receipt_ref  total_cost  total_price  \
0      3795898:230810:00400        70.2        161.1   
1      2209696:261008:02242        17.9         32.3   
2      3795898:251110:01261        64.2        153.0   
3      9626510:261028:02325        40.6         90.0   
4      3795898:260311:01405        84.0        192.0   
...                     ...         ...          ...   
13475  2209696:241203:01191        20.0         34.2   
13476  3795898:290411:02764        18.6         33.2   
13477  5593822:260202:01146         9.4         18.6   
13478  2209696:260227:01890        13.6         25.6   
13479  2209696:270427:02588        16.8         28.4   

       total_price_after_discount  total_profit  
0                          144.99         74.79  
1                           31.64         13.74  
2                          107.10         42.90  
3                           89.27         48.67  
4                          174.30         90.30  
...                           ...           ...  
13475                       34.20         14.20  
13476                       33.20         14.60  
13477                       18.60          9.20  
13478                       25.60         12.00  
13479                       28.40         11.60  

[13480 rows x 5 columns]

### Business Sales Profits - Joining all the the sales and their profits
Here we connecting the two previous tables to get an understanding of all the profits that made by the businesses

In [30]:
business_sales_profits_df

business_abn              business_name receipt_date  year  month  \
0        220 96 96        Wake Up with Coffee   2022/07/01  2022      7   
1        220 96 96        Wake Up with Coffee   2022/07/27  2022      7   
2        220 96 96        Wake Up with Coffee   2022/07/27  2022      7   
3        220 96 96        Wake Up with Coffee   2022/07/28  2022      7   
4        220 96 96        Wake Up with Coffee   2022/07/28  2022      7   
...            ...                        ...          ...   ...    ...   
13475    962 65 10  Please Bring Pizza Pronto   2029/06/13  2029      6   
13476    962 65 10  Please Bring Pizza Pronto   2029/06/14  2029      6   
13477    962 65 10  Please Bring Pizza Pronto   2029/06/14  2029      6   
13478    962 65 10  Please Bring Pizza Pronto   2029/06/14  2029      6   
13479    962 65 10  Please Bring Pizza Pronto   2029/06/29  2029      6   

      month_name           receipt_ref  receipt_total   gst  total_cost  \
0           July  2209696:220701:00002          60.50  5.50        36.3   
1           July  2209696:220727:00033          57.50  5.23        34.5   
2           July  2209696:220727:00034          64.00  5.82        38.4   
3           July  2209696:220728:00035          91.50  8.32        51.3   
4           July  2209696:220728:00036          40.00  3.64        24.0   
...          ...                   ...            ...   ...         ...   
13475       June  9626510:290613:04149          66.45  6.04        32.6   
13476       June  9626510:290614:04155          68.00  6.18        33.1   
13477       June  9626510:290614:04156          90.75  8.25        39.6   
13478       June  9626510:290614:04154         104.50  9.50        55.2   
13479       June  9626510:290629:04189          66.00  6.00        43.2   

       total_price  total_price_after_discount  total_profit  
0             60.5                       60.50         24.20  
1             57.5                       57.50         23.00  
2             64.0                       64.00         25.60  
3             91.5                       91.50         40.20  
4             40.0                       40.00         16.00  
...            ...                         ...           ...  
13475         71.0                       66.45         33.85  
13476         68.0                       68.00         34.90  
13477        104.4                       90.75         51.15  
13478        113.6                      104.50         49.30  
13479         66.0                       66.00         22.80  

[13480 rows x 13 columns]

In [31]:
with duckdb.connect(DUCKDB_FILE) as conn:
    business_monthly_profits_df = conn.sql("SELECT * FROM receipts.main.business_monthly_profits").df()


### Business Monthly Profit
Now that we know how much businesses are made by businesses, we cal aggregate how much profit do they make in each month.

In [33]:
business_monthly_profits_df["month-year"] = business_monthly_profits_df["month"].astype(str) + "-" + business_monthly_profits_df["year"].astype(str)
business_monthly_profits_df

business_abn         business_name  year  month month_name  \
0      379 58 98  Ed's Barber Supplies  2022      7       July   
1      379 58 98  Ed's Barber Supplies  2022      8     August   
2      379 58 98  Ed's Barber Supplies  2022      9  September   
3      379 58 98  Ed's Barber Supplies  2022     10    October   
4      379 58 98  Ed's Barber Supplies  2022     11   November   
..           ...                   ...   ...    ...        ...   
331    220 96 96   Wake Up with Coffee  2029      2   February   
332    220 96 96   Wake Up with Coffee  2029      3      March   
333    220 96 96   Wake Up with Coffee  2029      4      April   
334    220 96 96   Wake Up with Coffee  2029      5        May   
335    220 96 96   Wake Up with Coffee  2029      6       June   

     total_monthly_profit  num_of_sales month-year  
0                 1496.60            29     7-2022  
1                 1666.40            29     8-2022  
2                 1491.80            29     9-2022  
3                 1211.40            30    10-2022  
4                 1387.80            29    11-2022  
..                    ...           ...        ...  
331                815.69            58     2-2029  
332                782.62            57     3-2029  
333                762.94            55     4-2029  
334                778.57            55     5-2029  
335               1081.01            56     6-2029  

[336 rows x 8 columns]

### We want to visualise the performance of each business over the years

In [36]:
import plotly.graph_objects as go
businesses = business_monthly_profits_df["business_name"].unique()
business_colours = ["#70d6ff", "#ff70a6", "#ff9770", "#ffd670"]
fig = go.Figure()
for business, colour in zip(businesses, business_colours):
    business_profit_data = business_monthly_profits_df[
        business_monthly_profits_df["business_name"] == business]
    fig.add_trace(go.Scatter(x=business_profit_data["month-year"],
                             y=business_profit_data["total_monthly_profit"],
                             mode='lines+markers',
                             name=business,
                             line=dict(color=colour)
                             ))
fig.update_layout(
    title="Monthly Profits Across Businesses",
    xaxis_title="Months",
    yaxis_title="Total Profit ($)",
    xaxis_tickangle=270,
    legend_title="Businesses",
)
fig.show()

### Discussion

1. **Ed's Barber Supplies** is doing well year by year, and it is performing especially well during the first five months of the year.
![Ed's Barber Supplies](./images/q1_ed_barber_supplies.jpg)
2. **Please Bring Pizza Pronto** is performing really well across the year, and it is growing and doubling its profit every two years. However, it does look like its performance at the end of each year is slightly weaker than the rest of the year.
![Please Bring Pizza Pronto](./images/q1_please_bring_pizza.jpg)
3. **Penguin Swim School** performs worse in the middle of the year, which happens to be winter in Australia.  The business growth is sporadic and its overall trajectory is flat, which is to say, the business is not very stable and the growth is rather weak.
![Penguin Swim School](./images/q1_penguin_swim_school.jpg)
4. Wake Up with Coffee's business performance is weak with minimum growth compared to Ed's Barber Supplies and Please Bring Pizza Pronto,however, but it doesn't fluctuate as much as Penguin Swim School.  The business goes up and down but the pattern is not very clear and probably requires further investigation. 
![Wake Up With Coffee](./images/q1_wake_up_coffee.jpg)



---

## Exploration Two
### Q2 - Which customers were most loyal for each business?

### Introduction
Let's start by defining what do we mean by loyal customer.  Is loyal customer someone who bring in the most business?  Or is it someone who visit the business regularly or the most number of times.

### Methods

In [51]:
from typing import Literal
import plotly.graph_objects as go
from plotly.subplots import make_subplots

def show_bar_graph(receipts_df, header="Some Title", value_type: Literal["amount_spent", "num_purchases"]="amount_spent",):
    company_titles = [
        "Ed's Barber Supplies",
        'Penguin Swim School',
        'Wake Up with Coffee',
        'Please Bring Pizza Pronto'
    ]
    colours = ["#fb5607", "#ff006e", "#8338ec", "#3a86ff", ]
    fig = make_subplots(rows=2, cols=2, shared_yaxes=True, subplot_titles=company_titles,
                        vertical_spacing=0.35, row_heights=[0.8, 0.8,])

    business_name = company_titles[0]
    df = receipts_df[receipts_df['business_name'] == business_name]
    fig.add_trace(
        go.Bar(x=df.customer_name, y=df.amount_spent if value_type == "amount_spent" else df.num_purchases, text=df.amount_spent if value_type == "amount_spent" else df.num_purchases, textposition='outside', texttemplate='%{text:.2s}',
               name=business_name, marker=dict( color=colours[0])), row=1, col=1
    )

    business_name = company_titles[1]
    df = receipts_df[receipts_df['business_name'] == business_name]
    fig.add_trace(
        go.Bar(x=df.customer_name, y=df.amount_spent if value_type == "amount_spent" else df.num_purchases, text=df.amount_spent if value_type == "amount_spent" else df.num_purchases, textposition='outside', texttemplate='%{text:.2s}',
               name=business_name, marker=dict( color=colours[1])), row=1, col=2
    )

    business_name = company_titles[2]
    df = receipts_df[receipts_df['business_name'] == business_name]
    fig.add_trace(
        go.Bar(x=df.customer_name, y=df.amount_spent if value_type == "amount_spent" else df.num_purchases, text=df.amount_spent if value_type == "amount_spent" else df.num_purchases, textposition='outside', texttemplate='%{text:.2s}',
               name=business_name, marker=dict( color=colours[2])), row=2, col=1
    )

    business_name = company_titles[3]
    df = receipts_df[receipts_df['business_name'] == business_name]
    fig.add_trace(
        go.Bar(x=df.customer_name,
               y=df.amount_spent if value_type == "amount_spent" else df.num_purchases,
               text=df.amount_spent if value_type == "amount_spent" else df.num_purchases, textposition='outside', texttemplate='%{text:.2s}',
               name=business_name,
               marker=dict(color=colours[3]),
               ),
               row=2, col=2
    )
    fig.update_layout(title_text=header, height=600, showlegend=False)
    fig.show()


### Top 10 Customers by Amount Spent

In [52]:
with duckdb.connect(DUCKDB_FILE) as conn:
    top_ten_by_amount_spent_df = conn.sql("SELECT * FROM receipts.main.top_ten_customers_by_business_value").df()

top_ten_by_amount_spent_df

business_name customer_name  amount_spent  num_purchases  \
0        Ed's Barber Supplies          Lisa      12963.98            105   
1        Ed's Barber Supplies         Scott      10086.00             66   
2        Ed's Barber Supplies          Anna       8934.23             54   
3        Ed's Barber Supplies       Rebecca       8077.76             56   
4        Ed's Barber Supplies         Linda       7814.79             49   
5        Ed's Barber Supplies       William       7169.82             49   
6        Ed's Barber Supplies        Darren       7107.54             50   
7        Ed's Barber Supplies        Manuel       6933.37             53   
8        Ed's Barber Supplies         Carol       6932.46             54   
9        Ed's Barber Supplies        Mariah       6790.01             55   
10        Penguin Swim School         Scott       7570.42             96   
11        Penguin Swim School       William       7320.33             95   
12        Penguin Swim School      Samantha       5283.60             54   
13        Penguin Swim School         Derek       4682.62             57   
14        Penguin Swim School        Steven       4476.71             51   
15        Penguin Swim School       Natalie       4451.15             50   
16        Penguin Swim School     Johnathan       4356.51             55   
17        Penguin Swim School         Tammy       4245.23             51   
18        Penguin Swim School       Heather       4062.99             44   
19        Penguin Swim School        Meghan       4061.18             47   
20  Please Bring Pizza Pronto      Jonathan      10686.19            191   
21  Please Bring Pizza Pronto          Mary      10424.80            188   
22  Please Bring Pizza Pronto         Scott      10324.21            189   
23  Please Bring Pizza Pronto        Robert      10169.32            187   
24  Please Bring Pizza Pronto          Gary       9717.11            172   
25  Please Bring Pizza Pronto         Danny       9476.08            166   
26  Please Bring Pizza Pronto          Luke       9427.74            175   
27  Please Bring Pizza Pronto        Jeremy       9416.61            170   
28  Please Bring Pizza Pronto        Andrew       9128.57            159   
29  Please Bring Pizza Pronto         Aaron       8789.30            169   
30        Wake Up with Coffee         Scott       7098.17            192   
31        Wake Up with Coffee       Ricardo       4609.17            114   
32        Wake Up with Coffee        Robert       4596.26            120   
33        Wake Up with Coffee       Melissa       4336.75            112   
34        Wake Up with Coffee       Gregory       4236.98            113   
35        Wake Up with Coffee         Paula       4122.33            113   
36        Wake Up with Coffee        Justin       4064.74            111   
37        Wake Up with Coffee        Marvin       4045.93            104   
38        Wake Up with Coffee         Isaac       3950.81            111   
39        Wake Up with Coffee         Robin       3910.20            100   

    ranking  
0         1  
1         2  
2         3  
3         4  
4         5  
5         6  
6         7  
7         8  
8         9  
9        10  
10        1  
11        2  
12        3  
13        4  
14        5  
15        6  
16        7  
17        8  
18        9  
19       10  
20        1  
21        2  
22        3  
23        4  
24        5  
25        6  
26        7  
27        8  
28        9  
29       10  
30        1  
31        2  
32        3  
33        4  
34        5  
35        6  
36        7  
37        8  
38        9  
39       10

In [53]:
show_bar_graph(top_ten_by_amount_spent_df, header="Top Ten Customers by the Amount Spent")

#### Top 10 Customers by Number of Purchases

In [54]:
with duckdb.connect(DUCKDB_FILE) as conn:
    top_ten_by_num_purchases_df = conn.sql("SELECT * FROM receipts.main.top_ten_customers_by_num_purchases").df()

top_ten_by_num_purchases_df


business_name customer_name  num_purchases  amount_spent  \
0        Ed's Barber Supplies          Lisa            105      12963.98   
1        Ed's Barber Supplies         Scott             66      10086.00   
2        Ed's Barber Supplies       Rebecca             56       8077.76   
3        Ed's Barber Supplies        Mariah             55       6790.01   
4        Ed's Barber Supplies          Anna             54       8934.23   
5        Ed's Barber Supplies         Carol             54       6932.46   
6        Ed's Barber Supplies        Manuel             53       6933.37   
7        Ed's Barber Supplies        Sherry             51       6233.39   
8        Ed's Barber Supplies        Darren             50       7107.54   
9        Ed's Barber Supplies         David             50       6683.51   
10        Penguin Swim School         Scott             96       7570.42   
11        Penguin Swim School       William             95       7320.33   
12        Penguin Swim School         Derek             57       4682.62   
13        Penguin Swim School     Johnathan             55       4356.51   
14        Penguin Swim School      Samantha             54       5283.60   
15        Penguin Swim School          Cody             54       3443.07   
16        Penguin Swim School      Patricia             53       3715.63   
17        Penguin Swim School        Steven             51       4476.71   
18        Penguin Swim School         Tammy             51       4245.23   
19        Penguin Swim School       Natalie             50       4451.15   
20  Please Bring Pizza Pronto      Jonathan            191      10686.19   
21  Please Bring Pizza Pronto         Scott            189      10324.21   
22  Please Bring Pizza Pronto          Mary            188      10424.80   
23  Please Bring Pizza Pronto        Robert            187      10169.32   
24  Please Bring Pizza Pronto          Luke            175       9427.74   
25  Please Bring Pizza Pronto          Gary            172       9717.11   
26  Please Bring Pizza Pronto        Jeremy            170       9416.61   
27  Please Bring Pizza Pronto         Aaron            169       8789.30   
28  Please Bring Pizza Pronto        Steven            169       8441.17   
29  Please Bring Pizza Pronto         Danny            166       9476.08   
30        Wake Up with Coffee         Scott            192       7098.17   
31        Wake Up with Coffee        Robert            120       4596.26   
32        Wake Up with Coffee       Ricardo            114       4609.17   
33        Wake Up with Coffee       Gregory            113       4236.98   
34        Wake Up with Coffee         Paula            113       4122.33   
35        Wake Up with Coffee       Melissa            112       4336.75   
36        Wake Up with Coffee        Justin            111       4064.74   
37        Wake Up with Coffee         Isaac            111       3950.81   
38        Wake Up with Coffee        Leslie            106       3862.99   
39        Wake Up with Coffee        Marvin            104       4045.93   

    ranking  
0         1  
1         2  
2         3  
3         4  
4         5  
5         6  
6         7  
7         8  
8         9  
9        10  
10        1  
11        2  
12        3  
13        4  
14        5  
15        6  
16        7  
17        8  
18        9  
19       10  
20        1  
21        2  
22        3  
23        4  
24        5  
25        6  
26        7  
27        8  
28        9  
29       10  
30        1  
31        2  
32        3  
33        4  
34        5  
35        6  
36        7  
37        8  
38        9  
39       10

In [55]:
show_bar_graph(top_ten_by_num_purchases_df, header="Top 10 Customers by Number of Purchases", value_type="num_purchases")

### Discussion

## Top 10 Customers by Amount Spent

![Top 10 Customers by Amount Spent](./images/q2_top_10_customers_by_amount_spent.jpg)

## Top 10 Customers by Number of Purchases

![Top 10 Customers by Amount Spent](./images/q2_top_10_num_purchases.jpg)

---

## Exploration Three
### What is the employee turnover rate of each business?

### Introduction

Let's start by defining how to calculate employee turnover rate.  According to **nailed** website, employee turnover rate is calculated by dividing the number of employees left within a period, by the averaged number of employees at beginning and at the end of the period, and multiply by 100.

![Employee Turnover Rate](./images/etr.png))


I have also determined that a period is defined by the financial year, that is the year starts in July and ends in June.

1. "July 2022 - June 2023",
2. "July 2023 - June 2024",
3. "July 2024 - June 2025",
4. "July 2025 - June 2026",
5. "July 2026 - June 2027",
6. "July 2027 - June 2028",
7. "July 2028 - June 2029",

### Methods

In [57]:
periods = {
    1: "July 2022 - June 2023",
    2: "July 2023 - June 2024",
    3: "July 2024 - June 2025",
    4: "July 2025 - June 2026",
    5: "July 2026 - June 2027",
    6: "July 2027 - June 2028",
    7: "July 2028 - June 2029",
}

with duckdb.connect(DUCKDB_FILE) as conn:
    business_employees_df = conn.sql("""
    select * from business_employees
order by business_name, "year", "month", cashier_name
    """).df()

    business_employee_summary_df = conn.sql("""
    select business_name, year, month, month_name, num_employees from business_employee_stats
    """).df()


### Initial Employee Data

In [58]:
business_employees_df

business_name  year  month month_name cashier_name
0    Ed's Barber Supplies  2022      7       July      Anthony
1    Ed's Barber Supplies  2022      7       July     Jonathon
2    Ed's Barber Supplies  2022      8     August      Anthony
3    Ed's Barber Supplies  2022      8     August     Jonathon
4    Ed's Barber Supplies  2022      9  September      Anthony
..                    ...   ...    ...        ...          ...
860   Wake Up with Coffee  2029      6       June       Andrew
861   Wake Up with Coffee  2029      6       June        Grant
862   Wake Up with Coffee  2029      6       June       Hector
863   Wake Up with Coffee  2029      6       June     Patricia
864   Wake Up with Coffee  2029      6       June       Robert

[865 rows x 5 columns]

### Active Number of Employee Each Month

In [60]:
business_employee_summary_df

business_name  year  month month_name  num_employees
0    Ed's Barber Supplies  2022      7       July              2
1    Ed's Barber Supplies  2022      8     August              2
2    Ed's Barber Supplies  2022      9  September              2
3    Ed's Barber Supplies  2022     10    October              2
4    Ed's Barber Supplies  2022     11   November              2
..                    ...   ...    ...        ...            ...
331   Wake Up with Coffee  2029      2   February              2
332   Wake Up with Coffee  2029      3      March              2
333   Wake Up with Coffee  2029      4      April              2
334   Wake Up with Coffee  2029      5        May              3
335   Wake Up with Coffee  2029      6       June              5

[336 rows x 5 columns]

### Employee Leaving Stats
It is going to be faster for me to complete this step manually, that is to count the number of employee who left during each period.

In [66]:
business_employee_period_df = business_employee_summary_df[business_employee_summary_df["month"].isin([6, 7])].reset_index(drop=True)
business_employee_period_df["month-year"] = business_employee_period_df["month"].astype(str) + "-" +business_employee_period_df["year"].astype(str)

business_employee_period_df = business_employee_summary_df[business_employee_summary_df["month"].isin([6, 7])].reset_index(drop=True)
business_employee_period_df["month-year"] = business_employee_period_df["month"].astype(str) + "-" +business_employee_period_df["year"].astype(str)

businesses = ["Ed's Barber Supplies", 'Penguin Swim School', 'Please Bring Pizza Pronto', 'Wake Up with Coffee']


### Let's find out records of all the active employees/cashiers in each month.
From that we will deduce when the employee left.  The month that their names stop appearing on receipts, we will consider that they have left the business.

In [70]:
with duckdb.connect(DUCKDB_FILE) as conn:
    ed_employee_att_records_df = conn.sql(f"SELECT * FROM employee_attendences").df()

ed_employee_att_records_df


business_name  year  month month_name  \
0    Ed's Barber Supplies  2022      7       July   
1    Ed's Barber Supplies  2022      8     August   
2    Ed's Barber Supplies  2022      9  September   
3    Ed's Barber Supplies  2022     10    October   
4    Ed's Barber Supplies  2022     11   November   
..                    ...   ...    ...        ...   
331   Wake Up with Coffee  2029      2   February   
332   Wake Up with Coffee  2029      3      March   
333   Wake Up with Coffee  2029      4      April   
334   Wake Up with Coffee  2029      5        May   
335   Wake Up with Coffee  2029      6       June   

                            unique_employees  
0                          Anthony, Jonathon  
1                          Anthony, Jonathon  
2                          Anthony, Jonathon  
3                          Anthony, Jonathon  
4                          Anthony, Jonathon  
..                                       ...  
331                            Hector, Kelly  
332                            Hector, Kelly  
333                            Hector, Kelly  
334                    Brian, Hector, Lauren  
335  Andrew, Grant, Hector, Patricia, Robert  

[336 rows x 5 columns]

### Records when the employee left, based on the above table

In [71]:
# I have manually tallied the number of employees left during each period.
employee_left_records = {
    "Ed's Barber Supplies": {
        "July 2022 - June 2023": 0,
        "July 2023 - June 2024": 1, # Anthony
        "July 2024 - June 2025": 1, # Kathryn
        "July 2025 - June 2026": 1, # Jonathon
        "July 2026 - June 2027": 0, #
        "July 2027 - June 2028": 0, # Eric, Scott, Todd
        "July 2028 - June 2029": 11 # Natasha, Justin, Jennifer, Olivia, Michael, Juan, Barbara, Carlos, Robert, Cheryl, Ebony
        },
    "Penguin Swim School": {
        "July 2022 - June 2023": 0,
        "July 2023 - June 2024": 1, # Theresa
        "July 2024 - June 2025": 0, #
        "July 2025 - June 2026": 1, # Jennifer
        "July 2026 - June 2027": 0, #
        "July 2027 - June 2028": 0, #
        "July 2028 - June 2029": 4 # Kevin, Joseph, Mary, Patricia
        },
    "Please Bring Pizza Pronto": {
        "July 2022 - June 2023": 1, # Christian
        "July 2023 - June 2024": 1, # August
        "July 2024 - June 2025": 1, # Tara
        "July 2025 - June 2026": 0, #
        "July 2026 - June 2027": 2, # Ronald, Steven
        "July 2027 - June 2028": 1, # Maurice
        "July 2028 - June 2029": 1 # Amber
        },
    "Wake Up with Coffee": {
        "July 2022 - June 2023": 0, #
        "July 2023 - June 2024": 0, #
        "July 2024 - June 2025": 0, #
        "July 2025 - June 2026": 2, # Leslie, Amber
        "July 2026 - June 2027": 1, # Mary, Matthew
        "July 2027 - June 2028": 1, # Scott
        "July 2028 - June 2029": 6  # Diane, Nathaniel, Lauren, Kelly, Brian, Lauren
        },
    }


### Calculate the Employee Turnover Rate

In [73]:
final_stat = pd.DataFrame(columns=["business_name", "avg_turnover_rate", "median_turnover_rate",])

for business_name in businesses:
    ed_df = business_employee_period_df[business_employee_period_df["business_name"] == business_name].reset_index(drop=True)
    ed_df["period"] = (ed_df.index // 2) + 1
    ed_df["year_period"] = ed_df["period"].map(periods)
    ed_df = ed_df.pivot(index=["business_name", "year_period"], columns=["month_name"], values="num_employees")
    ed_df.reset_index(inplace=True)
    ed_df["annual_avg_employee"] = (ed_df["July"] + ed_df["June"])/2
    ed_df["num_employee_left"] = 0

    business_employee_attendances = ed_employee_att_records_df[ed_employee_att_records_df["business_name"] == business_name]
    # st.dataframe(business_employee_attendances)
    ed_df["num_employee_left"] = ed_df["year_period"].map(employee_left_records[business_name])
    ed_df["annual_turnover_rate"] = (ed_df["num_employee_left"] / ed_df["annual_avg_employee"]) * 100

    final_stat.loc[len(final_stat)] = [business_name, ed_df["annual_turnover_rate"].mean(), ed_df["annual_turnover_rate"].median()]
    
final_stat["avg_turnover_rate"] = final_stat["avg_turnover_rate"].round(2)
final_stat

business_name  avg_turnover_rate  median_turnover_rate
0       Ed's Barber Supplies              62.04                  40.0
1        Penguin Swim School              38.37                   0.0
2  Please Bring Pizza Pronto              36.26                  40.0
3        Wake Up with Coffee              46.87                  40.0

In [74]:
colours = ["#fb5607", "#ff006e",]
data = [go.Bar(name="Avg. Turnover Rate", x=final_stat["business_name"], y=final_stat["avg_turnover_rate"],
               text=final_stat["avg_turnover_rate"], marker=dict(color=colours[0])),
        go.Bar(name="Med. Turnover Rate", x=final_stat["business_name"], y=final_stat["median_turnover_rate"],
               text=final_stat["median_turnover_rate"], marker=dict(color=colours[1])),]
fig = go.Figure(data=data)
fig.update_layout(barmode="group")
fig.show()

### Discussion

Employee turnover rate is a crucial metric for businesses to understand the stability of their workforce.

It is rather depressing to look at the table above, but we have to take into consideration that these are very 
small businesses with 1-5 employees at any given time. Losing just one or two employees a year will cause the 
employee turnover rate to spike. In general, these businesses may lose 0-2 employees a year. However, for some 
reason, during the period from July 2028 to July 2029, these businesses churned through employees, which caused 
the employee turnover rate to increase suddenly and skew the result.

If you look at Penguin Swim School for example, its averaged employee turnover rate is 38.37% but it's median turnover rate is 0%.  That is because most year they hardly loose any employee, but then on the last financial year, they **maybe** made some bad hiring decisions and went through a lot more number of employees and thus caused the averaged turnover rate to increase sharply.

The averaged employee turnover rates of the businesses ares:
1. Ed's Barber Supplies - 62.04%
2. Penguin Swim School - 38.37%
3. Please Bring Pizza Pronto - 36.26%
4. Wake Up with Coffee - 46.87%

The median employee turnover rates of the businesses are:
1. Ed's Barber Supplies - 40%
2. Penguin Swim School - 0%
3. Please Bring Pizza Pronto - 40%
4. Wake Up with Coffee - 40%